packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,257 @@
+import cdk = require('@aws-cdk/cdk');
+import { PolicyStatement } from "./policy-document";
+import { IPrincipal } from "./principals";
+
+/**
+ * Basic options for a grant operation
+ */
+export interface CommonGrantOptions {
+  /**
+   * The principal to grant to
+   *
+   * @default No work is done
+   */
+  principal?: IPrincipal;
+
+  /**
+   * The actions to grant
+   */
+  actions: string[];
+
+  /**
+   * The resource ARNs to grant to
+   */
+  resourceArns: string[];
+}
+
+/**
+ * Options for a grant operation
+ */
+export interface GrantWithResourceOptions extends CommonGrantOptions {
+  /**
+   * The resource with a resource policy
+   *
+   * The statement will be added to the resource policy if it couldn't be
+   * added to the principal policy.
+   */
+  resource: IResourceWithPolicy;
+
+  /**
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+/**
+ * Options for a grant operation that only applies to principals
+ */
+export interface GrantOnPrincipalOptions extends CommonGrantOptions {
+  /**
+   * Construct to report warnings on in case grant could not be registered
+   */
+  scope?: cdk.IConstruct;
+}
+
+/**
+ * Options for a grant operation to both identity and resource
+ */
+export interface GrantOnPrincipalAndResourceOptions extends CommonGrantOptions {
```

Comment:
looks exactly like `GrantWithResourceOptions`...

Reply:
Correct, but that's not necessarily true forever and Java users will have to type this class name, so we can't tie them together or it would be a breaking change in the future if one of them needed a new field.